File: backend/schemas/booking.py
from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, Field, model_validator

BookingStatus = Literal["pending", "accepted", "rejected", "cancelled", "completed"]
PaymentStatus = Literal["unpaid", "paid_cash", "paid_card", "paid_paypal"]
PaymentMethod = Literal["cash", "credit_card", "paypal"]


class BookingCreate(BaseModel):
    provider_id: int
    package_id: int | None = None
    wedding_package_id: int | None = None
    event_date: datetime
    guest_count: int | None = None
    phone_number: str | None = None
    location: str = Field(min_length=2, max_length=255)
    notes: str | None = None
    payment_method: PaymentMethod

    @model_validator(mode="after")
    def validate_future_event(self) -> "BookingCreate":
        event_date = self.event_date
        now = datetime.now(timezone.utc) if event_date.tzinfo else datetime.now()
        compare_date = event_date.astimezone(timezone.utc) if event_date.tzinfo else event_date
        if compare_date <= now:
            raise ValueError("Event date and time must be in the future.")
        return self


class BookingStatusUpdate(BaseModel):
    booking_status: BookingStatus
    provider_response_note: str | None = Field(default=None, max_length=1000)


class BookingProviderResponse(BaseModel):
    provider_response_note: str | None = Field(default=None, max_length=1000)


class BookingAdminUpdate(BaseModel):
    booking_status: BookingStatus | None = None
    payment_status: PaymentStatus | None = None


class BookingOut(BaseModel):
    id: int
    customer_id: int
    provider_id: int | None = None
    package_id: int | None = None
    wedding_package_id: int | None = None
    customer_name: str | None = None
    provider_name: str | None = None
    package_title: str | None = None
    wedding_package_title: str | None = None
    event_date: datetime | None = None
    guest_count: int | None = None
    phone_number: str | None = None
    location: str | None = None
    notes: str | None = None
    total_price: float
    booking_status: BookingStatus
    provider_response_note: str | None = None
    customer_confirmed: bool
    provider_confirmed: bool
    conversation_id: int | None = None
    payment_method: str | None = None
    payment_status: str
    created_at: datetime
    updated_at: datetime

File: frontend/src/types/booking.ts
export type BookingStatus = "pending" | "accepted" | "rejected" | "cancelled" | "completed";
export type PaymentMethod = "cash" | "credit_card" | "paypal";
export type PaymentStatus = "unpaid" | "paid_cash" | "paid_card" | "paid_paypal";

export type BookingSeed = {
  providerId: number;
  providerName: string;
  packageId?: number;
  packageTitle?: string;
  weddingPackageId?: number;
  totalPrice?: number;
};

export type BookingFormData = {
  event_date: string;
  event_time: string;
  guest_count: string;
  phone_number: string;
  location: string;
  notes: string;
};

export type BookingPayload = {
  provider_id: number;
  package_id?: number;
  wedding_package_id?: number;
  event_date: string;
  guest_count?: number;
  phone_number?: string;
  location: string;
  notes?: string;
  payment_method: PaymentMethod;
};

export type Booking = {
  id: number;
  customer_id: number;
  provider_id?: number | null;
  package_id?: number | null;
  wedding_package_id?: number | null;
  customer_name?: string | null;
  provider_name?: string | null;
  package_title?: string | null;
  wedding_package_title?: string | null;
  event_date?: string | null;
  guest_count?: number | null;
  phone_number?: string | null;
  location?: string | null;
  notes?: string | null;
  total_price: number;
  booking_status: BookingStatus;
  provider_response_note?: string | null;
  customer_confirmed: boolean;
  provider_confirmed: boolean;
  conversation_id?: number | null;
  payment_method?: PaymentMethod | null;
  payment_status: PaymentStatus;
  created_at: string;
  updated_at: string;
};
